### PR TITLE
Set container env var using services.

### PR DESCRIPTION
**Pull request by yashdesai93**
This change introduces setting container env var using services.
It also adds tests for the same.

### CODE REVIEW

File: vkubelet/env.go [hunk 1]
```diff
@@ -347,6 +400,29 @@ loop:
 			continue loop
 		}
 	}
+
+	// enableServiceLinks defaults to true.
+	enableServiceLinks := true
+	if pod.Spec.EnableServiceLinks != nil {
```
Comment (cpuguy83) [inline vkubelet/env.go]: The kubelet errors out if this is nil, I think we should do the same.

File: vkubelet/env.go [hunk 2]
```diff
@@ -89,6 +95,53 @@ func populateContainerEnvironment(ctx context.Context, pod *corev1.Pod, containe
...
+	for i := range services {
+		service := services[i]
+		// ignore services where ClusterIP is "None" or empty
+		if !v1helper.IsServiceIPSet(service) {
+			continue
+		}
+		serviceName := service.Name
+
+		// We always want to add environment variabled for master services
+		// from the master service namespace, even if enableServiceLinks is false.
```
Comment (cpuguy83) [inline vkubelet/env.go]: I understand this is basically a copy/paste from kubelet, but can we come up with a better than "master"?

File: vkubelet/env.go [hunk 3]
```diff
@@ -89,6 +95,53 @@ func populateContainerEnvironment(ctx context.Context, pod *corev1.Pod, containe
...
+		if service.Namespace == metav1.NamespaceDefault && masterServices.Has(serviceName) {
+			if _, exists := serviceMap[serviceName]; !exists {
+				serviceMap[serviceName] = service
+			}
+		} else if service.Namespace == ns && enableServiceLinks {
+			serviceMap[serviceName] = service
+		}
+	}
+
+	mappedServices := []*corev1.Service{}
```
Comment (cpuguy83) [inline vkubelet/env.go]:
```suggestion
	mappedServices := make([]*corev1.Service, 0, len(serviceMap))
```

File: vkubelet/env.go (hunk 1 above)
Comment (yashdesai93) [reply vkubelet/env.go]: My only concern was that some e2e tests might fail, but let's see.

**Comment by yashdesai93** [comment]
@cpuguy83 failing `pod.Spec.EnableServiceLinks` being nil led to the timeout. I haven't figured out the exact cause, but I have added a TODO and for now use the default when nil. I am hoping that's okay.

**Pull request merged**